### PR TITLE
[jit] Make classtypes hold a weak_ptr to their CU

### PR DESCRIPTION
**Pull request by suo**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #22727 [jit] Make `load()` create only one CU
* **#22726 [jit] Make classtypes hold a weak_ptr to their CU**
* #22725 [jit] Make traced fns also go into the global python CU
* #22724 [jit] _script_compile and _script_class_compile add to the python CU

This breaks the ownership cycle between modules and their functions.

gh-metadata: pytorch pytorch 22726 gh/suo/91/head

Differential Revision: [D16197608](https://our.internmc.facebook.com/intern/diff/D16197608)

### CODE REVIEW

File: t [hunk 1]
```diff
@@ -0,0 +1,58 @@
+[1mdiff --cc torch/csrc/jit/import_source.cpp[m
```
Comment (suo) [inline t]: what in the world

File: torch/csrc/jit/script/class_type.cpp [hunk 2]
```diff
@@ -9,19 +9,21 @@ namespace c10 {
...
 Function* ClassType::getMethod(const std::string& name) const {
   const auto qualname = QualifiedName(*qualified_name_obj(), name);
-  return compilation_unit_->find_function(qualname);
+  auto cu = compilation_unit_.lock();
+  TORCH_INTERNAL_ASSERT(cu);
+  return cu->find_function(qualname);
 }
 
-std::shared_ptr<CompilationUnit> ClassType::compilation_unit() {
+std::weak_ptr<CompilationUnit> ClassType::compilation_unit() {
```
Comment (zdevito) [inline torch/csrc/jit/script/class_type.cpp]: I think this should return a strong pointer via lock(). weak_ptr is an implementation detail of the class.

File: torch/csrc/jit/register_prim_ops.cpp [hunk 3]
```diff
@@ -1074,8 +1074,10 @@ RegisterOperators reg(
            const auto type = node->output()->type()->expect<ClassType>();
            const size_t numAttrs = type->numAttributes();
            return [type, numAttrs](Stack& stack) {
+             auto cu = type->compilation_unit().lock();
```
Comment (zdevito) [inline torch/csrc/jit/register_prim_ops.cpp]: StrongTypePtr and the cu lookup should go _outside_ the closure to prevent recompute every time.

**Review by zdevito — APPROVED** [review]
Looks good. minor nits

**Comment by bddppq** [comment]
This has broken CUDA CI on master:
```
Jul 15 21:32:01 ======================================================================
Jul 15 21:32:01 FAIL: test_lerp_cuda (__main__.TestFuser)
Jul 15 21:32:01 ----------------------------------------------------------------------
Jul 15 21:32:01 Traceback (most recent call last):
Jul 15 21:32:01   File "/var/lib/jenkins/workspace/test/common_utils.py", line 399, in wrapper
Jul 15 21:32:01     method(*args, **kwargs)
Jul 15 21:32:01   File "/var/lib/jenkins/workspace/test/common_utils.py", line 344, in __exit__
Jul 15 21:32:01     self.name, after - before, i))
Jul 15 21:32:01   File "/var/lib/jenkins/workspace/test/common_utils.py", line 578, in assertEqual
Jul 15 21:32:01     super(TestCase, self).assertLessEqual(abs(x - y), prec, message)
Jul 15 21:32:01 AssertionError: 512 not less than or equal to 1e-05 : __main__.TestFuser.test_lerp_cuda leaked 512 bytes CUDA memory on device 0
Jul 15 21:32:01 
Jul 15 21:32:01 ----------------------------------------------------------------------
Jul 15 21:32:01 Ran 46 tests in 28.454s
Jul 15 21:32:01 
Jul 15 21:32:01 FAILED (failures=1, skipped=2)
Jul 15 21:32:02 Traceback (most recent call last):
Jul 15 21:32:02   File "test/run_test.py", line 434, in <module>
Jul 15 21:32:02     main()
Jul 15 21:32:02   File "test/run_test.py", line 426, in main
Jul 15 21:32:02     raise RuntimeError(message)
Jul 15 21:32:02 RuntimeError: test_jit_fuser failed!
```

**Pull request closed without merging**